### ISSUE TITLE
Adjacency & Laplacian matrices do not reflect edge weights

### ISSUE THREAD

**Issue by BoundaryValueProblems** [sbromberger/SimpleWeightedGraphs.jl#37]
I constructed a weighted graph using ```SimpleWeightedGraphs.jl``` and tried to get its adjacency and Laplacian matrices via ```adjacency_matrix(G)``` and ```laplacian_matrix(G)```. However, these matrices are the same as those of the unweighted case (i.e., all the edges have the same weight 1). It would be great if these matrices reflect the edge weights assigned. Or do I need to use some other functions to get proper versions of these matrices? Thanks for your help! Here is my MWE:
```Julia
# creating a simple unweighted path
G=SimpleGraph(3);
add_edge!(G, 1, 2);
add_edge!(G, 2, 3);
W=Array(adjacency_matrix(G))
3×3 Array{Int64,2}:
 0  1  0
 1  0  1
 0  1  0
L=Array(laplacian_matrix(G))
3×3 Array{Int64,2}:
  1  -1   0
 -1   2  -1
  0  -1   1
# now for a weighted path graph
Gw=SimpleWeightedGraph(3);
add_edge!(Gw, 1, 2, 5.0);
add_edge!(Gw, 2, 3, 10.0);
Ww=Array(adjacency_matrix(Gw))
3×3 Array{Int64,2}:
 0  1  0
 1  0  1
 0  1  0
Lw=Array(laplacian_matrix(Gw))
3×3 Array{Int64,2}:
 1  -1   0
-1   2  -1
 0  -1   1
```
Of course, I want to have the last two matrices as follows:
```Julia
Ww
3×3 Array{Float64,2}:
 0.0  5.0  0.0
 5.0  0.0 10.0
 0.0 10.0  0.0
Lw
3×3 Array{Float64,2}:
 5.0  -5.0   0.0
-5.0  15.0 -10.0
 0.0 -10.0  10.0
```

**Comment by sbromberger** [comment]
This is what `weights()` is for.

**Comment by BoundaryValueProblems** [comment]
@sbromberger, thanks! OK, then ```weights(Gw)``` extracts a proper weighted adjacency matrix of ```Gw``` that is a weighted graph in my above MWE. But still I have to do:
```
Lw = Diagonal(sum(Ww,dims=1)[:])-Ww
```
to construct the corresponding Laplacian matrix. In my opinion, it would be better to have ```laplacian_matrix``` and ```adjacency_matrix``` for a ```SimpleWeightedGraph``` object, which returns the proper matrices.

**Comment by sbromberger** [comment]
Can you point me to any literature that takes edge weights into account for either adjacency or laplacian matrices? I haven't seen any in a cursory search, and it doesn't make sense to me that you'd use edge weights here, especially since
```
julia> g = Graph(3)
{3, 0} undirected simple Int64 graph

julia> add_edge!(g, 1, 1)
true

julia> adjacency_matrix(g)
3×3 SparseMatrixCSC{Int64,Int64} with 1 stored entry:
  [1, 1]  =  2
```

**Comment by BoundaryValueProblems** [comment]
@sbromberger ; Normally, the Laplacian matrix ```L``` of a graph ```G``` regardless of weighted or unweighted is defined as ```L=D-W``` where ```D``` is the degree matrix (diagonal) and ```W``` is the adjacency matrix of ```G```.  If ```G``` is a weighted graph, then ```W[i,j]```  is the edge weight between ```i```-th and ```j```-th vertices.  And ```D[i,i]``` is the sum of the ```i```-th row (or column) of ```W```, which represents the weighted degree of the ```i```-th vertex. This definition is a standard one in the spectral graph theory. You can take a look at various books on the spectral graph theory, algebraic graph theory, or data analysis (e.g., C. Godsil & G. Royle: _Algebraic Graph Theory_, 2001, Springer; D. Skillicorn: _Understnading Complex Datasets: Data Mining with Matrix Decompositions_, Chapman & Hall/CRC, 2007; but I also cite the following popular paper:
Ulrike von Luxburg: A tutorial on spectral clustering, _Statistical Computing_, vol.17, pp.395–416,  2007. DOI 10.1007/s11222-007-9033-z

**Comment by sbromberger** [comment]
> If `G` is a weighted graph, then `W[i,j]` is the edge weight between `I`-th and `j`-th vertices

I have not seen this referenced in any material yet. In fact, in most theory that I've seen, `W` is not the adjacency matrix, it is the **weight** matrix (hence `W` instead of `A`). Therefore, the weighted Laplacian uses `D-W`, which should simply be `D-weights(g)` for `AbstractSimpleWeightedGraph`.

If you wanted to add a `weighted_laplacian` function to SWG, I'd ask @jpfairbanks to review it but I wouldn't necessarily be opposed. Right now, however, I think I'd like to keep the existing functions as they are, since they are pretty well understood.

**Comment by BoundaryValueProblems** [comment]
@sbromberger; Well, although I do not want to dwell on the terminology issues too much, there are many published articles (e.g., those appear in the journal _Linear Algebra and its Application_) dealing with graph Laplacian matrices of weighted graphs that use this definition of the (weighted) adjacency matrix. Often people call the weight matrix as a weighted adjacency matrix or simply an adjacency matrix. Then, the graph Laplacian matrix of a weighted graph is sometimes called the weighted Laplacian matrix, but more often simply called the Laplacian matrix. 
For example, in addition to the paper of von Luxburg I mentioned earlier, you can take a look at:
* R.B. Bapat, D. Kalita, S. Pati: On weighted directed graphs, _Linear Algebra and its Applications_, Volume 436, Issue 1, 1 January 2012, Pages 99-111. https://doi.org/10.1016/j.laa.2011.06.035
* T. Bıyıko ̆glu, J. Leydold: Graphs of given order and size and minimum algebraic connectivity,  _Linear Algebra and its Applications_, Volume 436, Issue 7, 1 April 2012, Pages 2067-2077. https://doi.org/10.1016/j.laa.2011.09.026

Anyways, to avoid possible confusion, it is a good idea to define ```weighted_laplacian``` or ```weighted_laplacian_matrix``` function for ```AbstractSimpleWeightedGraph```. Since there is a function ```laplacian_matrix``` for a ```Graph``` object in ```LightGraphs```, why not? Also, it may be a good idea to define ```weighted_adjacency_matrix``` function as a synonym of ```weights```. @jpfairbanks , could you consider these?

**Comment by jpfairbanks** [comment]
Yeah I agree that weighted laplacian is commonly called laplacian when the weighted context is obvious. Since LightGraphs generally treats weights as a secondary property of graphs unless you are using WeightedGraphs, I think using `weighted_laplacian` would make sense for AbstractGraph, but `laplacian(g::SimpleWeightedGraph) = diag(sum(diag(weights(g))) - weights(g) .* adjacency_matrix(g)` would make sense as a default for SimpleWeightedGraph.

You could also use the LaplacianMatrix type from the LinAlg module something like `LinAlg.LaplacianMatrix(weights(A) .* adjacency_matrix(A))` should work. This will create a type that stores the weighted adjacency matrix and the sum of the weights that behaves like the weighted Laplacian for the purposes of matrix multiplication, which is probably what you need.

**Comment by yuehhua** [comment]
I agree with this. Providing some implementations in python networkx package:

The features of [adjacency_matrix](https://networkx.github.io/documentation/stable/reference/generated/networkx.linalg.graphmatrix.adjacency_matrix.html#networkx.linalg.graphmatrix.adjacency_matrix) and [laplacian_matrix](https://networkx.github.io/documentation/stable/reference/generated/networkx.linalg.laplacianmatrix.laplacian_matrix.html#networkx.linalg.laplacianmatrix.laplacian_matrix) provide the option to use edge weight to calculate adjacency matrix and laplacian matrix.

Statement in the doc:
> weight (string or None, optional (default=’weight’)) – The edge data key used to compute each value in the matrix. If None, then each edge has weight 1.

And I need this feature...

**Comment by sbromberger** [comment]
@yuehhua a PR would be appreciated. It seems fairly straightforward to override `laplacian_matrix` using @jpfairbanks' code above.

**Comment by yuehhua** [comment]
OK, I take this PR. #48

**Comment by yuehhua** [comment]
Solved by #48.

**Comment by sbromberger** [comment]
Closed via #48.
